### DISCUSSION TITLE
jbrowse add-track color definition

### DISCUSSION THREAD

**Discussion by OnlineArts** [GMOD/jbrowse-components#2700] [Q&A] [answered]
Hi,

I tried to find an answer, but the only related issue I found was GMOD/jbrowse#4 from 2012 and probably not actual anymore.

How to define a specific color either by setting a category or via a config string through jbrowse add-track.

I tried:
```bash
/usr/bin/jbrowse add-track /opt/mosga/snakemake/../gui/uploads/3e5bbcd1/genome.gff.gz --out /opt/mosga/snakemake/../gui/uploads/3e5bbcd1/jbrowse --load copy --category=Annotation --subDir=Genes --force -n MOSGA --config='{"style":{"color":"red"}}'
```
or
```bash
/usr/bin/jbrowse add-track /opt/mosga/snakemake/../gui/uploads/3e5bbcd1/genome.gff.gz --out /opt/mosga/snakemake/../gui/uploads/3e5bbcd1/jbrowse --load copy --category=Annotation --subDir=Genes --force -n MOSGA --config='{"color":"red"}'
```
Both have no effect. Is it possible to do it like that or even better, by the category ("Annotation" in this case)?
 

- jbrowse/cli/1.2.0 linux-x64 node-v14.18.3
- Firefox 96

**Comment by garrettjstevens** [comment, marked as answer]
Hi, I've transferred this issue to the JBrowse 2 repository, since it looks like that's what you're using based on the use of the JBrowse 2 CLI. If that's not correct, please let me know.

You can change the color of a GFF3 track by using a `--config` like this one:

```
--config '{"displays":[{"type":"LinearBasicDisplay","displayId":"myDisplayId","renderer":{"type":"SvgFeatureRenderer","color1":"red"}}]}'
```

(See a similar answer [here](https://github.com/GMOD/jbrowse-components/discussions/1487#discussioncomment-277009))

There are a couple things going on here. The `displays` refers to how this track is being displayed (e.g. in a linear view, a circular view, etc.), and we customize the linear display here. The renderer option `color1` is the main color for the feature, but there are some other colors you can customize. Check out the options under "renderer" on the right-hand side area [here](https://s3.amazonaws.com/jbrowse.org/code/jb2/main/index.html?config=test_data%2Fvolvox%2Fconfig.json&session=share-gzvxZebgZ4&password=NnRIZ) for what else is available.

This would be a good thing to document, but we are also working on updating our feature rendering, so those docs might not get written until after that's done.

**Comment by OnlineArts** [comment]
Thank you @garrettjstevens !
I looked in the documentation but was not able to resolve it. Although you plan to restructure that, may you consider allowing a bypass parameter like ``--color #f00``.

Your example worked fine for me until using it for multiple tracks in the same session. I recieved the follow error code:
```Error: [mobx-state-tree] Cannot resolve a reference to type 'LinearBasicDisplayConfigurationSchema' with id: 'myDisplayId' unambigously, there are multiple candidates: /jbrowse/tracks/0/displays/0, /jbrowse/tracks/1/displays/0, /jbrowse/tracks/2/displays/0, /jbrowse/tracks/3/displays/0, /jbrowse/tracks/4/displays/0```

I assume that comes from the "displayId": "myDisplayId", right? It could be even more a reason to think about an easy possibility to allow individual colors because I think without setting the displays-config manually, it would define a corresponding displayId without conflicts. :-)

**Reply by garrettjstevens** [reply]
Yes, that comes from the "displayId". Each track you define will have to have a unique "displayId".

One problem with a bypass like `--color` is that it would have to be interpreted differently for GFF3 than for e.g. BAM or BigWig tracks. Also, that wouldn't provide any information about if that color was supposed to apply in a linear view or a circular view, etc. We do need some sort of simplification, but it might take us some time to design something that will work across different types of views and displays.

**Comment by OnlineArts** [comment]
I'm total with you.
I suspected that the working solution is the most liberal but lacks a bit of accessibility. Obviously, the config parameter allows further configuration like value-dependent colors, etc. Bypasses should not hinder the config parameter. I hope you will find a simplified parameter that covers the most expectation context-depending (GFF/BAM/linear/circular etc).

Thank you for your support! It worked for my use-case.
![JBrowse](https://user-images.githubusercontent.com/17415804/152137855-71a6f7f9-1012-45f7-ab79-d5366e3e978e.png)
